xbmc/interfaces/generic/RunningScriptObserver.cpp

```diff
@@ -0,0 +1,37 @@
+/*
+ *  Copyright (C) 2017-2021 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "RunningScriptObserver.h"
+
+#include "interfaces/generic/ScriptInvocationManager.h"
+
+CRunningScriptObserver::CRunningScriptObserver(int scriptId, CEvent& evt)
+  : m_scriptId(scriptId),
+  m_event(evt),
+  m_thread(this, "ScriptObs"),
+  m_stopEvent(true)
+{
+  m_thread.Create();
+}
+
+void CRunningScriptObserver::Run()
+{
+  do
+  {
+    if (!CScriptInvocationManager::GetInstance().IsRunning(m_scriptId))
+    {
+      m_event.Set();
+      break;
+    }
+  } while (!m_stopEvent.WaitMSec(20));
+}
+
+void CRunningScriptObserver::Abort()
+{
+  m_stopEvent.Set();
```

Comment:
If you replace the line with 'm_thread.StopThread ();', there are no more crashes on Android.

I think that on Android, there might be a wrong thread terminating and this causing Kodi to crash.

Reply:
Thanks for the input. I think I understand why this should solve the problem. `CRunningScriptObserver` is created in `CScriptRunner::WaitOnScriptResult()` as a local variable. Right before the local variable goes out of scope it calls `CRunningScriptObserver::Abort()`. I assumed that the destructor of `CRunningScriptObserver` would implicitly stop the thread because it would also execute the destructor of `m_thread` but depending on the order of destruction `m_stopEvent` is destroyed before `m_thread` so when the next loop in `CRunningScriptObserver::Run()` is executed `m_stopEvent` has become invalid.

The question is whether calling `CRunningScriptObserver::Abort()` should be blocking or not. If it should I need to add a `m_thread.StopThread()` after the `m_stopEvent.Set()`. If it shouldn't block I need to add a destructor to `CRunningScriptObserver` which first calls `Abort()` and then `m_thread.StopThread()`. Looking at the original code of `CPluginDirectory::CScriptObserver::Abort()` it also called `StopThread()` so that's probably the proper solution.